Format plugin for AS/400 SHA-1 hashes

rschoem opened the pull request:
Format plugin for AS/400 SHA-1 hashes (**'as400-sha'**)
See http://hackthelegacy.org for more info and additional tooling.
Initial version by **@5up3rUs3r** and **@bartholozz**.

frank-dittrich commented on src/as400_sha_fmt_plug.c:
```diff
@@ -0,0 +1,419 @@
+{
+	// Format of each line: {"hash like in internal representation", "valid password for hash", {"userid as in file","hash as in file"}}
+	{"ROB       4C106E52CA196986E1C52C7FCD02AF046B76C73C", "banaan", {"ROB","4C106E52CA196986E1C52C7FCD02AF046B76C73C"}},
+	{"BART      CED8050C275A5005D101051FF5BCCADF693E8AB7", "Kulach007", {"BART","CED8050C275A5005D101051FF5BCCADF693E8AB7"}},
```
For john, I think we need a format specific prefix added.

Are the user names always upper case?
Are non-ascii characters allowed? If yes, a sample hash would be good. If not, valid() might be more restrictive.

magnumripper replied:
I agree we need a tag for pot files. We should be able to load hashes without it though.

frank-dittrich commented on src/as400_sha_fmt_plug.c:
```diff
@@ -0,0 +1,419 @@
+	}
+};
+
+#endif /* plugin stanza */
```
Does your editor support EditorConfig? If yes, it is easier to avoid the missing newline.

magnumripper commented on src/as400_sha_fmt_plug.c:
```diff
@@ -0,0 +1,419 @@
+// Structures and global variables
+
+// Charset used for testing if ciphertext string is hexadecimal
+static const char AS400_BASE16_CHARSET[] = "0123456789ABCDEFabcdef";
```
Should we allow lower-case hex? If hashes "come" in uppercase there's no reason really.

magnumripper commented on src/as400_sha_fmt_plug.c:
```diff
@@ -0,0 +1,419 @@
+#define BINARY_LENGTH         20                                 // Binary length of hashes
+#define BINARY_ALIGN          1
+#define SALT_SIZE             sizeof(struct custom_salt)
+#define SALT_ALIGN            4
```
I believe the salt align should be 1 as the get_salt() is currently written.

magnumripper commented on src/as400_sha_fmt_plug.c:
```diff
@@ -0,0 +1,419 @@
+	{
+		password_utf16be[(i*2)]=0x00;
+		password_utf16be[(i*2)+1]=password[i];
+	}
```
This is not a proper UTF-16 conversion unless input is ISO-8859-1. We have functions that do it better, see eg. oracle_fmt_plug.c

magnumripper commented on src/as400_sha_fmt_plug.c:
```diff
@@ -0,0 +1,419 @@
+		ALGORITHM_NAME,
+		BENCHMARK_COMMENT,
+		BENCHMARK_LENGTH,
+        0,
```
Bad indentation here. All indentation must be tabs (alignment with spaces).

magnumripper commented:
The format should be rewritten to use a salt and a binary. It's currently written as a non-hash format and that only harms performance.

magnumripper commented:
It fails self-test on the build bots.

jfoug commented:
Does this format have any EBCDIC issues that we will have to address with our encoding?

magnumripper commented:
Doesn't look like it. It only mentions UTF-16BE.

kholia commented:
http://www.hackthelegacy.org/attachments/pwd400gen.py

EBCDIC is not involved in AS/400 SHA-1 hashing. However it is used in "IBM DES" hashing.

rschoem commented:
Guys, thank you all for the comments.
Made some modifications:
- fixed bad indentation;
- SALT_ALIGN = 1
- Missing newline
- Fixed a few bad memory allocations.
Tested on Windows (Cygwin), Kali Linux, Ubuntu.
Will close this pull request and resubmit new version.
Will look into UTF16 conversion and rewrite with binary and salt later.
